pkg/R/serialize.R

```diff
@@ -0,0 +1,121 @@
+# Utility functions to serialize R objects so they can be read in Java.
+
+# Type mapping from R to Java
+#  
+# integer -> Int
+# character -> String
+# logical -> Boolean
+# double, numeric -> Double
+# raw -> Array[Byte]
+#
+# list[T] -> Array[T], where T is one of above mentioned types
+# environment -> Map[String, T], where T is a native type
+# jobj -> Object, where jobj is an object created in the backend
+
+writeObject <- function(con, object, withType = TRUE) {
+  # NOTE: In R vectors have same type as objects. So we don't support
+  # passing in vectors as arrays and instead require arrays to be passed
+  # as lists.
+  if (withType) {
+    writeType(con, class(object))
+  }
+  switch(class(object),
+    integer = writeInt(con, object),
+    character = writeString(con, object),
+    logical = writeBoolean(con, object),
+    double = writeDouble(con, object),
+    numeric = writeDouble(con, object),
+    raw = writeRaw(con, object),
+    list = writeList(con, object),
+    jobj = writeString(con, object$id),
+    environment = writeEnv(con, object),
+    stop("Unsupported type for serialization"))
+}
+
+writeString <- function(con, value) {
+  writeInt(con, as.integer(nchar(value) + 1))
```

Comment:
Yeah - so the string value in R contains a `\0` at the end (like strings in say C) and writeBin writes it out